server/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java

```diff
@@ -43,14 +44,16 @@
  * In particular, this policy will delete index commits whose max sequence number is at most
  * the current global checkpoint except the index commit which has the highest max sequence number among those.
  */
-public final class CombinedDeletionPolicy extends IndexDeletionPolicy {
+public class CombinedDeletionPolicy extends IndexDeletionPolicy {
     private final Logger logger;
     private final TranslogDeletionPolicy translogDeletionPolicy;
     private final SoftDeletesPolicy softDeletesPolicy;
     private final LongSupplier globalCheckpointSupplier;
     private final ObjectIntHashMap<IndexCommit> snapshottedCommits; // Number of snapshots held against each commit point.
     private volatile IndexCommit safeCommit; // the most recent safe commit point - its max_seqno at most the persisted global checkpoint.
     private volatile IndexCommit lastCommit; // the most recent commit point
+    private volatile SafeCommitInfo safeCommitInfo = SafeCommitInfo.EMPTY;
+    private final Object onCommitMutex = new Object();
```

Comment:
Do we really need this separate mutex?

Reply:
I am not sure, because it depends on whether `onCommit` is called concurrently or not. Is it? If it is then we could end up with stale data in `safeCommitInfo`.

Reply:
Before this change, onCommit was a synchronized method. Is it okay if we leave it as is?

Reply:
Henning asked to avoid running the IO needed to compute the `safeCommitInfo` under this mutex here: https://github.com/elastic/elasticsearch/pull/45208#discussion_r310984670.

Reply:
Ah, ok. Although It's not an issue, I think we should avoid having two lock orderings: 
`this -> onCommitMutex` in `onInit` and `onCommitMutex -> this` in `onCommit`. How about implementing `onCommit` in two steps using a single mutex.

```java
final IndexCommit safeCommit;
synchronized (this) {
	// update the policy
        safeCommit = this.safeCommit;
}
final SafeCommitInfo safeCommitInfo = ..
synchronized (this) {
	if (safeCommit == this.safeCommit) {
		this.safeCommitInfo = safeCommitInfo;
	}
}
```